src/aviary/functional.py

```diff
@@ -0,0 +1,173 @@
+from collections.abc import Callable
+from functools import update_wrapper
+from typing import Any
+
+from pydantic import BaseModel, Field
+
+from aviary.env import Environment, Frame
+from aviary.message import Message
+from aviary.tools import Messages, Tool, ToolRequestMessage
+from aviary.utils import is_coroutine_callable
+
+
+class DynamicState(BaseModel):
+    """Dynamic env state model that adapts to provided extras."""
+
+    reward: float = 0
+    done: bool = False
+    extras: dict[str, Any] = Field(default_factory=dict)
+
+    def __getattr__(self, name: str):
+        """Allow direct access to extras as attributes."""
+        if name in self.extras:
+            return self.extras[name]
+        raise AttributeError(f"'State' has no attribute {name!r}")
+
+
+class FunctionalEnvironment(Environment[DynamicState]):
+    """Environment class for function-based environments."""
+
+    def __init__(
+        self,
+        start_fn: Callable,
+        tools: list[Tool],
+        allow_concurrency: bool,
+        *args,
+        **kwargs,
+    ):
+        super().__init__()
+        self.start_fn = start_fn
+        self.tools = tools
+        self.args = args
+        self.kwargs = kwargs
+        self.state = DynamicState()
+        self.allow_concurrency = allow_concurrency
+
+    async def _call_start(self) -> tuple[str, dict[str, Any]]:
+        """Call start_fn handling both sync and async cases."""
+        if is_coroutine_callable(self.start_fn):
+            result = await self.start_fn(*self.args, **self.kwargs)
+        else:
+            result = self.start_fn(*self.args, **self.kwargs)
+
+        if not isinstance(result, tuple) or len(result) != 2:  # noqa: PLR2004
+            raise TypeError("Start function must return (observation, state_dict)")
+
+        obs, state_dict = result
+        if not isinstance(state_dict, dict):
+            raise TypeError("State must be a dictionary")
+
+        return obs, state_dict
+
+    async def reset(self) -> tuple[Messages, list[Tool]]:
+        obs, extras = await self._call_start()
+        self.state = DynamicState(extras=extras)
+        return [Message(content=obs)], self.tools
```

Comment:
```suggestion
        obs, state_dict = result
        if not isinstance(state_dict, dict):
            raise TypeError("State must be a dictionary")

        return obs, DynamicState(extras=state_dict)

    async def reset(self) -> tuple[Messages, list[Tool]]:
        obs, self.state = await self._call_start()
        return [Message(content=obs)], self.tools
```

Why not go ahead and convert to `DynamicState` in `_call_start`? Is there any value in deferring the `DynamicState` construction to `reset`?

Reply:
no reason, guess we can switch.